Enable running integration tests in the finagle-redis project.

mkhq opened the pull request:
Problem

Unit tests and integration tests are mixed in the finagle-redis project under the test path. In addition, the SBT configuration is only running unit tests. This partially solves issue #360.

Solution

Moved the integration test files from `test` to `it`. Extracted traits used for integration tests and created a new file under `it` with these traits. Added configuration for IntegrationTest in the finagle-redis project build.

Result

Running `test` only executes unit tests. Running `it:test` executes the integration tests.

mosesn commented:
BtreeClient is something twitter-specific, I believe.  I'm not sure where else Btrees in redis exist, and I don't think anyone at twitter uses finagle-redis anymore.  Maybe we should just delete the btree support?  If anyone wants to revive it, it will still be in the git commit history.

mkhq commented:
Hey, thanks for taking a look at this. I removed the Btree integration tests and all tests should work now.

@mosesn, may I ask if you are not using finagle-redis, how are you accessing redis from scala?

mosesn commented:
We have a cache proxy which we speak thrift to instead.  I don't know the details / rationale, I haven't used it personally.

dschobel commented on finagle-redis/src/it/scala/com/twitter/finagle/redis/RedisIntegrationTest.scala:
```diff
@@ -0,0 +1,89 @@
+import org.jboss.netty.buffer.ChannelBuffer
+import org.scalatest.{BeforeAndAfterAll, FunSuite}
+import java.net.InetSocketAddress
+import com.twitter.finagle.Service
```
sort please

dschobel commented on finagle-redis/src/it/scala/com/twitter/finagle/redis/RedisIntegrationTest.scala:
```diff
@@ -0,0 +1,89 @@
+
+  protected val OKStatusReply = StatusReply("OK")
+
+  protected def withRedisClient(testCode: Service[Command, Reply] => Any) {
```
: Unit =

dschobel commented on finagle-redis/src/it/scala/com/twitter/finagle/redis/RedisIntegrationTest.scala:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  protected def assertMBulkReply(reply: Future[Reply], expects: List[String],
+    contains: Boolean = false) = Await.result(reply) match {
```
can you break this method up w/ one param per line as in:

def lotsOfArgs(
  number: Option[String],
  typeOfPhone: String
  hoursOfAvailability: Range = 9.to(12)
)

dschobel commented on finagle-redis/src/it/scala/com/twitter/finagle/redis/RedisIntegrationTest.scala:
```diff
@@ -0,0 +1,89 @@
+        case true =>
+          assert(expects.isEmpty === false, "Test did no supply a list of expected replies.")
+          val newMsgs = ReplyFormat.toString(msgs)
+          expects.foreach({ msg =>
```
we can just pass the block to foreach without the parens

dschobel commented on finagle-redis/src/it/scala/com/twitter/finagle/redis/RedisIntegrationTest.scala:
```diff
@@ -0,0 +1,89 @@
+            assert(doesMBulkReplyContainMessage === true)
+          })
+        case false =>
+          val actualMessages = ReplyFormat.toChannelBuffers(msgs).map({ msg =>
```
no parens for map

dschobel commented:
just style nits, LGTM, merges cleanly and integration tests pass

penland365 commented:
@dschobel In fairness to @mkhq , most of those style nits were introduced by me last year when I was still trying to get a handle on writing proper Scala.

mosesn commented:
OK, I'm going to try to fix up the style nits myself so we can get this in (since this has taken so long).  Thanks @mkhq!

mkhq commented:
@mosesn sorry didnt fix above earlier, should be in the commit now. There are some conflicts with develop now, will take a look.

mkhq commented:
@mosesn Merged in develop and updated a couple of things to the latest code base, e.g. PubSub was contributed with two integration test suites. Several of the tests are failing when running `it:test`, so trying to solve that.

mosesn commented:
@mkhq thanks for the update!  Let me know if there's anything I can do to help.

mkhq commented:
@mosesn, Submitted another commit, there are still a couple of unresolved issues:

mosesn commented:
@mkhq do you mind holding off on working on this for a little while?  we're doing a major refactor of finagle-redis, and I'm worried about your work accidentally getting lost in it.  it should land in ~ a week.

mkhq commented:
@mosesn sure, no problem. The current state of this is 7 failing tests, mostly due to NPE instead of ClientError.

mosesn commented:
@mkhq thanks for waiting!  We've made the changes to redis, so it should be in a pretty reasonable state now.  Thanks for your patience! 🚀

mkhq commented:
@mosesn merged in latest develop, all tests should work now after removing the Server related tests. A short summary:

- Run integration tests using `it:test`
- Tested using Redis 3.2
- Removed btree and topology tests since this is not part of the official release
- Fixed scans by replacing PATTERN with MATCH
- Updated the expected DUMP results, seems this differs between versions.
- Fixed a bug when running sentinel tests, not all servers shutdown correctly
- Fixed sentinel protocol to use link-pending-commands instead of pending-commands

If this looks ok, should I squash the commits?

mosesn commented:
Yeah, squashing would be rad.  I can do a code review when you feel like it's ready 👍

mkhq commented:
@mosesn Feel free to look at the code. Squashing is a bit messy since I've merged in develop a couple of times. One alternative to fix the history is to create a new PR and close this one. Do you have any suggestions?

mosesn commented:
A new PR works for me, thanks so much!

mkhq commented:
@mosesn Created a new PR https://github.com/twitter/finagle/pull/543

nepthar commented:
New PR - #543. Closing this.
